### DISCUSSION TITLE
Planet import taking too long on placex data loading

### DISCUSSION THREAD

**Discussion by micaelcapitao** [osm-search/Nominatim#3335] [General]
Hi folks,

I'm new to Nominatim and I'm currently trying to do a full Planet import. However, the import process is on the placex table loading for 5 days already. I can see I have a few cores completely maxed to while the actual disk writes are on the kB/s order.

At this point I'm out of ideas and given each attempt takes multiple days, randomly trying is very unproductive :(

Let me share with you my setup and where the process is currently at.

Do you have any ideas on how to debug or speed up this process? Any hints would be much appreciated :)

Thank you!

# Environment

The whole process is running on GCP VMs: one for the postgres database, another for the Nominatim import process.

Postgres VM:
- 20 vCPUs
- 128 GB memory
- OS: Ubuntu 20.04.6 LTS
- Data disk: SSD, 2 TB

Postgres details:
- PostgreSQL 15.5 (Ubuntu 15.5-1.pgdg20.04+1)
- POSTGIS="3.4.1 ca035b9" [EXTENSION] PGSQL="150" GEOS="3.8.0-CAPI-1.13.1 " PROJ="6.3.1" GDAL="GDAL 3.0.4, released 2020/01/28" LIBXML="2.9.10" LIBJSON="0.13.1" LIBPROTOBUF="1.3.3" WAGYU="0.5.0 (Internal)" RASTER
- Settings (the ones I adjusted):
```
max_connections: 200
shared_buffers: 32GB
effective_cache_size: 96GB
checkpoint_completion_target: 0.9
wal_buffers: 16MB
default_statistics_target: 100
random_page_cost: 1.1
effective_io_concurrency: 200
#work_mem: 41943kB
huge_pages: try
min_wal_size: 1GB
max_wal_size: 4GB
max_worker_processes: 20
max_parallel_workers_per_gather: 4
max_parallel_workers: 20
max_parallel_maintenance_workers: 4
statement_timeout: 0
maintenance_work_mem: 10GB
autovacuum_work_mem: 2GB
work_mem: 50MB
synchronous_commit: off
checkpoint_timeout: 10min
```

Nominatim import VM:
- 8 vCPUs
- 64 GB memory
- OS: Ubuntu 22.04.3 LTS
- Disk: SSD, 200 GB
- Nominatim version: 4.3.2

# Process

I've used the following Nominatim options:
```
NOMINATIM_FLATNODE_FILE=/path/to/flatnode
NOMINATIM_OSM_FILE=https://planet.openstreetmap.org/pbf/planet-latest.osm.pbf
```

In terms of extra data I'm including these in the project directory:
```
country_grid.sql.gz
gb_postcodes.csv.gz
us_postcodes.csv.gz
wikimedia-importance.sql.gz
```

And, finally, I'm running this command:
```
nominatim import --threads 8 --osm2pgsql-cache 16384
```

# Status

From the log of `nominatim import` I can see the following:
```
2024-02-06 18:13:15: Using project directory: /app/nominatim-planet
2024-02-06 18:13:18: Creating database
2024-02-06 18:13:19: Setting up country tables
2024-02-06 18:13:21: Importing OSM data file
2024-02-06 18:13:21  osm2pgsql version 1.9.2
2024-02-06 18:13:21  Database version: 15.5 (Ubuntu 15.5-1.pgdg20.04+1)
2024-02-06 18:13:21  PostGIS version: 3.4
2024-02-06 18:13:21  Storing properties to table '"public"."osm2pgsql_properties"'.
...
2024-02-07 06:20:55    Processed 8910995322 nodes in 6557s (1h 49m 17s) - 1359k/s
2024-02-07 06:20:55    Processed 996773447 ways in 30347s (8h 25m 47s) - 33k/s
2024-02-07 06:20:55    Processed 11804581 relations in 6750s (1h 52m 30s) - 2k/s
2024-02-07 06:20:55  No marked ways (Skipping stage 2).
2024-02-07 06:20:55  Clustering table 'place' by geometry...
2024-02-07 07:06:49  No indexes to create on table 'place'.
2024-02-07 07:06:49  Creating id index on table 'place'...
2024-02-07 07:18:53  Analyzing table 'place'...
2024-02-07 07:18:54  Done postprocessing on table 'planet_osm_nodes' in 0s
2024-02-07 07:18:54  Building index on table 'planet_osm_ways'
2024-02-07 10:41:04  Done postprocessing on table 'planet_osm_ways' in 12129s (3h 22m 9s)
2024-02-07 10:41:04  Building index on table 'planet_osm_rels'
2024-02-07 10:45:40  Done postprocessing on table 'planet_osm_rels' in 276s (4m 36s)
2024-02-07 10:45:40  All postprocessing on table 'place' done in 3479s (57m 59s).
2024-02-07 10:45:40  Storing properties to table '"public"."osm2pgsql_properties"'.
2024-02-07 10:45:40  osm2pgsql took 59539s (16h 32m 19s) overall.
2024-02-07 10:45:41: Importing wikipedia importance data
2024-02-07 10:47:18: Importing secondary importance raster data
2024-02-07 10:47:18: Secondary importance file not imported. Falling back to default ranking.
2024-02-07 10:47:18: Create functions (1st pass)
2024-02-07 10:47:19: Create tables
2024-02-07 10:54:42: Create functions (2nd pass)
2024-02-07 10:54:42: Create table triggers
2024-02-07 10:54:42: Create partition tables
2024-02-07 10:54:48: Create functions (3rd pass)
2024-02-07 10:54:48: Initialise tables
2024-02-07 10:54:51: Load data into placex table
...
<since then it is still dealing with this placex table>
```

The VM running the `nominatim import` is currently completely idle just waiting for postgres.

On the postgres side I can see it is running 7 queries that look like:
```
 INSERT INTO placex ("osm_type","osm_id","class","type","name","admin_level","address","extratags","geometry")                        
                            SELECT "osm_type","osm_id","class","type","name","admin_level","address","extratags","geometry" FROM place
                            WHERE osm_id % 7 = 0                                                                                      
                              AND NOT (class='place' and (type='houses' or type='postcode'))                                          
                              AND ST_IsValid(geometry)
```
7 CPUs of this VM are at 100%. In terms of IO the writes are at about 100 kB/s.
Currently the data disk has 412 GB occupied.

**Comment by mtmail** [comment]
During this stage of the import seeing all CPU cores at maximum, for several days, is expected. What are the last lines of the import log? Which rank does it process?
In the database you can check for deadlocks but we haven't had any reports after Nominatim switched off the `jit` Postgresql setting (in the session, no change to the .conf files needed).

**Reply by micaelcapitao** [reply]
Hi,

Thanks for replying.
From what I've read so far, I was expecting the ranking process to take a long time. The problem is that it hasn't event reached the ranking stage apparently.

The last lines of the log are:
```
2024-02-07 10:54:51: Load data into placex table
.......................................................................................................................................
```
And the `...` are just being appended until now. Nothing else.

I'm not sure about the best way to check for deadlocks. Please let me know if this is good enough:
```
=# select pid, 
       usename, 
       pg_blocking_pids(pid) as blocked_by, 
       query as blocked_query
from pg_stat_activity
where cardinality(pg_blocking_pids(pid)) > 0;

 pid | usename | blocked_by | blocked_query
-----+---------+------------+---------------
(0 rows)
```
Apparently no deadlocks.

**Reply by mtmail** [reply]
I see, so it's still in the loading phase, hasn't started indexing yet. Deadlocks wouldn't happen here. Actually taking days wouldn't happen either, I'd expect 1-2 hours.
What's unusual is loading into an external database, most users use a local database. Is the database server on a volume that needs to expand? Bascially maybe the operating system or hoster tries to save "real" disc space and only expands the volume on demand. I'm not familiar with Google Cloud but I've seen it happen with Vmware. If it's not a configuration setting a trick could be to write a huge file (dd utility), then delete it before the import.
Lastly it could be network between database server and its storage, we'd had reports (again VMware) where that was an issue. With 100kb/s and high CPU that seems unlikely. Even a HDD on the other side of the world would have better load performance.

**Reply by micaelcapitao** [reply]
Let me give that a check.
Do you happen to have a full postgres config for a big machine that you know it worked? That way I could try to find something off on my postgres config.
Knowing that the `placex` loading should take like 1-2 hours is already a good hint. Pretty much all this time that I'm now waiting would be expected for the ranking... and not for this `placex` loading.
Thank you!

**Comment by geoneutrino** [comment]
This sounds really slow
Short ago i did a planet import on a bare metal similar to your VM with 16core/64gb ram/debian with postgres15 and import took ~30 hours, placex step 52 minutes

I would point to the "2 system" scenario or a generic slowness either storage or network (is there traffic ?)

What is 
NOMINATIM_FLATNODE_FILE=/path/to/flatnode
is the path on the import machine or network share ?
I ask because your osm2pgsql takes 16.5 hours, on my machine it took only 7.5 hours and all stages are much faster even the nodes -  your "Processed 8910995322 nodes" (flatnodes file) took 110 minutes, mine only 33 minutes
So even with a factor 3 or 4 in performance your placex should finish in a couple of hours

Does syslog (both machines) or postgres log show something ?

Can you test with import and DB on one VM to narrow this down ?

**Reply by micaelcapitao** [reply]
Hi,

> is there traffic ?

At this stage, no, no traffic between the 2 machines. The placex thing is just a few INSERT queries as I mentioned.

> NOMINATIM_FLATNODE_FILE=/path/to/flatnode
is the path on the import machine or network share ?

It is a path on the import machine, on an SSD.

> I ask because your osm2pgsql takes 16.5 hours

How many threads did you use for your import? And did you have to adjust the `osm2pgsql-cache`?

> Does syslog (both machines) or postgres log show something ?

Nothing that does seems like an error/warning or anything :|

**Reply by geoneutrino** [reply]
i didnt use threads and osm2pgsql-cache param

postgres i used the params for a "64 gb ram machine" from https://nominatim.org/release-docs/latest/admin/Installation/
but i also turned fsync and autovacuum off (dangerous, dont forget to turn back on after import !!) this speeds up a bit but not by factors

**Reply by micaelcapitao** [reply]
I have both `fsync` and `autovacuum` enabled. Might be worth to try disabling them!
I also have `maintenance_work_mem: 10GB` but just noticed my postgres is using `autovacuum_work_mem: -1` when it should be `autovacuum_work_mem: 2GB` as per my config. Not sure the impact of this but worth checking!

Thanks for the hints. Having a better understanding of the time scale helps already.

I'll keep you posted!

**Reply by lonvia** [reply]
`fsync` makes little difference for me these days, so I usually leave it on. That might be because of NVMEs though. There might be a difference on a VM.

Do not disable autovacuum. It does an important job of updating statistics in the background.

**Reply by geoneutrino** [reply]
@lonvia i started an import and after ~20 hours it was at Rank 26 with ETA (of this rank) > 100.000 seconds
This seemed bad compared to the last time i imported a planet (> 1 year ago i think) so i deactivated fsync/autovaccum as i did last times and this resulted in the 30hour import. After turning on both again the autovacuum process did lots of I/O in the background as expected for a longer time to update statistics but the advantage was to have the system running and continue setup/tests earlier

Yes, its an NVMe Storage, 4 discs as Raid-10, this makes an impact compared to a VM but 100kb/s as he measured don't point directly to i/o overload

**Reply by micaelcapitao** [reply]
As @lonvia pointed out, my `placex_insert` trigger didn't look as expected for the import phase and I believe I [found the reason for that](https://github.com/osm-search/Nominatim/discussions/3335#discussioncomment-8484011).
Let's see how it behaves now. I don't need it to be super fast, as long as it finishes, eventually...

**Comment by lonvia** [comment]
The stage you are in is really just a copy from one (huuge) table to another. There is an INSERT trigger involved but it shouldn't be doing very complicated things. If it has been hanging for 5 days though, I suspect that something in the trigger is going wrong. This is really tricky to debug.

First of all, can you check which triggers are installed on the `placex` table? (Do `\d placex` in the psql command line)
Then, please check the content of the placex_insert trigger (`\df+ placex_insert`). Does the function definition contain the code at https://github.com/osm-search/Nominatim/blob/master/lib-sql/functions/placex_triggers.sql#L686-L737

**Reply by micaelcapitao** [reply]
Hi,
So these are the triggers:
```
Triggers:
    placex_before_delete AFTER DELETE ON placex FOR EACH ROW EXECUTE FUNCTION placex_delete()
    placex_before_insert BEFORE INSERT ON placex FOR EACH ROW EXECUTE FUNCTION placex_insert()
    placex_before_update BEFORE UPDATE ON placex FOR EACH ROW EXECUTE FUNCTION placex_update()
```

And checking the `placex_insert ()` contents I can see they are exactly the same on that section you highlighted.

**Reply by lonvia** [reply]
Interesting, this section shouldn't be in there during import. Here is what the function should look like:

```nominatim=# \df+ placex_insert
                                                                                                                     List of functions
 Schema |     Name      | Result data type | Argument data types | Type | Volatility | Parallel | Owner  | Security | Access privileges | Language |                                       Source code                                       | Description 
--------+---------------+------------------+---------------------+------+------------+----------+--------+----------+-------------------+----------+-----------------------------------------------------------------------------------------+-------------
 public | placex_insert | trigger          |                     | func | volatile   | unsafe   | suzuki | invoker  |                   | plpgsql  |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          | DECLARE                                                                                +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   postcode TEXT;                                                                       +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   result BOOLEAN;                                                                      +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   is_area BOOLEAN;                                                                     +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   country_code VARCHAR(2);                                                             +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   diameter FLOAT;                                                                      +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   classtable TEXT;                                                                     +| 
        |               |                  |                     |      |            |          |        |          |                   |          | BEGIN                                                                                  +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   NEW.place_id := nextval('seq_place');                                                +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   NEW.indexed_status := 1; --STATUS_NEW                                                +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   NEW.centroid := ST_PointOnSurface(NEW.geometry);                                     +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   NEW.country_code := lower(get_country_code(NEW.centroid));                           +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   NEW.partition := get_partition(NEW.country_code);                                    +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   NEW.geometry_sector := geometry_sector(NEW.partition, NEW.centroid);                 +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   IF NEW.osm_type = 'X' THEN                                                           +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     -- E'X'ternal records should already be in the right format so do nothing          +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   ELSE                                                                                 +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     is_area := ST_GeometryType(NEW.geometry) IN ('ST_Polygon','ST_MultiPolygon');      +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     IF NEW.class in ('place','boundary')                                               +| 
        |               |                  |                     |      |            |          |        |          |                   |          |        AND NEW.type in ('postcode','postal_code')                                      +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     THEN                                                                               +| 
        |               |                  |                     |      |            |          |        |          |                   |          |       IF NEW.address IS NULL OR NOT NEW.address ? 'postcode' THEN                      +| 
        |               |                  |                     |      |            |          |        |          |                   |          |           -- most likely just a part of a multipolygon postcode boundary, throw it away+| 
        |               |                  |                     |      |            |          |        |          |                   |          |           RETURN NULL;                                                                 +| 
        |               |                  |                     |      |            |          |        |          |                   |          |       END IF;                                                                          +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |       NEW.name := hstore('ref', NEW.address->'postcode');                              +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     ELSEIF NEW.class = 'highway' AND is_area AND NEW.name is null                      +| 
        |               |                  |                     |      |            |          |        |          |                   |          |            AND NEW.extratags ? 'area' AND NEW.extratags->'area' = 'yes'                +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     THEN                                                                               +| 
        |               |                  |                     |      |            |          |        |          |                   |          |         RETURN NULL;                                                                   +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     ELSEIF NEW.class = 'boundary' AND NOT is_area                                      +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     THEN                                                                               +| 
        |               |                  |                     |      |            |          |        |          |                   |          |         RETURN NULL;                                                                   +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     ELSEIF NEW.class = 'boundary' AND NEW.type = 'administrative'                      +| 
        |               |                  |                     |      |            |          |        |          |                   |          |            AND NEW.admin_level <= 4 AND NEW.osm_type = 'W'                             +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     THEN                                                                               +| 
        |               |                  |                     |      |            |          |        |          |                   |          |         RETURN NULL;                                                                   +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     END IF;                                                                            +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     SELECT * INTO NEW.rank_search, NEW.rank_address                                    +| 
        |               |                  |                     |      |            |          |        |          |                   |          |       FROM compute_place_rank(NEW.country_code,                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                               CASE WHEN is_area THEN 'A' ELSE NEW.osm_type END,        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                               NEW.class, NEW.type, NEW.admin_level,                    +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                               (NEW.extratags->'capital') = 'yes',                      +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                               NEW.address->'postcode');                                +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     -- a country code make no sense below rank 4 (country)                             +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     IF NEW.rank_search < 4 THEN                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |       NEW.country_code := NULL;                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |     END IF;                                                                            +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   END IF;                                                                              +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |  -- not disable_diff_updates                                                           +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          |   RETURN NEW;                                                                          +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                        +| 
        |               |                  |                     |      |            |          |        |          |                   |          | END;                                                                                   +| 
        |               |                  |                     |      |            |          |        |          |                   |          |                                                                                         | 
(1 row)
```

If your function is longer and contains the highlighted section, you can cancel your import now. It will never finish.

Have you done something unusual likes stopping and restarting the import?

**Reply by lonvia** [reply]
To restart the process:
* Run `nominatim refresh --functions --no-diff-updates`
* Reconfirm with `\df+ placex_insert` that the highlighted section is gone and the function now looks as above.
* Continue import with: `nominatim import --continue load-data`

**Reply by micaelcapitao** [reply]
> Have you done something unusual likes stopping and restarting the import?

I have not. My import script looks like this:
```
export NOMINATIM_DATABASE_DSN="pgsql:dbname=nominatim;host=${DB_HOST:?};port=${DB_PORT:?};user=nominatim;password=${DB_PASS:?}"
export NOMINATIM_FLATNODE_FILE="${DATA_DIR:?}/flatnode"

wget https://nominatim.org/data/country_grid.sql.gz -O country_grid.sql.gz
wget https://nominatim.org/data/gb_postcodes.csv.gz -O gb_postcodes.csv.gz
wget https://nominatim.org/data/us_postcodes.csv.gz -O us_postcodes.csv.gz
wget https://nominatim.org/data/wikimedia-importance.sql.gz -O wikimedia-importance.sql.gz

dropdb -h "${DB_HOST}" -p "${DB_PORT}" -U nominatim --if-exists --force nominatim

nominatim import --threads 8 --osm2pgsql-cache 16384   # CURRENTLY STUCK HERE!!!

wget https://nominatim.org/data/tiger-nominatim-preprocessed-latest.csv.tar.gz -O tiger.csv.tar.gz
nominatim add-data --tiger-data tiger.csv.tar.gz

nominatim index --threads 8
nominatim admin --check-database
nominatim admin --warm
```

Every time I attempt to do the import, a drop database always happens before so I'd expect to always start clean...

**Reply by micaelcapitao** [reply]
I believe I found the culprit.
There was an old deployment of Nominatim, pointing to this same database, that was constantly running a `nominatim replication --init`. That should explain the trigger not being the expected one at import time.
Testing a clean import now and I'll let you know the results.

**Comment by micaelcapitao** [comment]
Good morning folks,

After a few hiccups (some infra related issues causing my VMs to reboot), everything went well.
Total time around 88 hours:
- 27 hours from beginning to finish placex loading.
  - placex loading was around 4 hours.
- Then 26 hours to finish rank 26.
- Another 26 hours to finish rank 30 - this time here is a bit of a guess because in the middle of it the VMs where rebooted: I then restarted the process with `--continue indexing`.
- 1.5 hours for post-process tables.
- 7 hours for the recompute word counts.

For a next trial, I believe it will be a good idea to give more CPUs to the machine used for the loading (and reduce the database side as well). Everything was being done with a parallelism of 8. The database side was definitely underused all the time.

Thank you for the help! :)

**Reply by lonvia** [reply]
Sounds about right for the setup.

I would expect that you don't even need to change the specification of the loader machine. Simply run the import with `-j20` to match the parallelism of the database machine. Most of the import work for Nominatim happens in the database, so that should be the point of reference. If you do test that, I'd be interested to hear about the results.

**Reply by micaelcapitao** [reply]
Initially, as I pointed in the beginning of the discussion, I was setting the number of threads (`--threads`) but then I ran into memory issues during the `osm2pgsql` stage which made to adjust the cache size and finally I tried not setting anything at all and it worked. For a next time (should be soon enough actually) I'll try with a higher number of threads and see how it goes.

**Comment by pkthogit** [comment]
Hi Team,

Yesterday I have kicked off import process and found myself also into the same situation where I see it taking too long time, almost more than 6hr at **"Load data into placex table...."** (osm2pgsql took 48419s (13h 26m 59s) overall.)

> I believe I found the culprit. There was an old deployment of Nominatim, pointing to this same database, that was constantly running a `nominatim replication --init`. That should explain the trigger not being the expected one at import time. Testing a clean import now and I'll let you know the results.

But in my case, I have got a brand new cloud VM to start installing for the first time. However, I notice, based on entire conversation in this thread, that Im not using machine resources properly. It would be great if someone could review my machine details and what is the best way to fully utilize the resources to get he installation of nominatim done at the earliest. (especially postgres parameter, import command parameters also cleaning up database to start fresh import)

**Environment:**

I have installed both postgres database and nominatim on the same cloud vm, where postgres database data directory is pointed to another mounted drive having 5TB space.

370 GB RAM memory
48 threads
OS: Ubuntu 22.04.3 LTS
Disk: SSD, 4TB
Nominatim version: 4.3.2
Postgres 16


**Postgres database parameters:**
shared_buffers = 100GB
work_mem = 1GB
maintenance_work_mem = 10GB
effective_cache_size = 24GB
synchronous_commit = off
max_wal_size = 1GB
checkpoint_timeout = 10min
checkpoint_completion_target = 0.9
fsync = off
full_page_writes = off

**Commands executed:**
cd /srv/nominatim/build/
wget https://www.nominatim.org/data/wikimedia-importance.sql.gz
wget https://www.nominatim.org/data/us_postcode_data.sql.gz
wget https://www.nominatim.org/data/gb_postcode_data.sql.gz
wget -O country_osm_grid.sql.gz https://www.nominatim.org/data/country_grid.sql.gz

/usr/local/bin/nominatim import -j 6 --osm-file /srv/nominatim/build/planet-latest.osm.pbf 2>&1 | tee setup.log

-Prem

**Reply by mtmail** [reply]
(Adding a new question as an answer to an older closed discussion will lead to confusion)

I've seen "osm2pgsql took" with 7 hours on good non-VM hardware. The biggest factor is if you use a flatnode file, for a planet import you should (https://nominatim.org/release-docs/latest/admin/Import/#flatnode-files). Another factor is which data was selected, the so called style (https://nominatim.org/release-docs/latest/admin/Import/#filtering-imported-data) which can make comparing one import to another harder.

osm2pgsql is single threaded at this step so what counts is raw CPU processor power plus how fast IO is. That's harder to determine with cloud providers, many don't mention processor clock speeds. If you have the option use the fastest IO, I think at AWS that's measured in IOPS (higher = faster) and ideally the IO should be on the machine not network mounted.

The next steps "Starting rank 3" are multithreaded. I'd increase the `-j 6` to at least 20. That will likely make the full import twice as fast on your hardware. If you have time to test (with smaller data, not the whole planet) try 30 or 40 even.
